File: arc061/many_formulas.rb
s = gets.chomp
@ans = 0
def dfs(s,sum,count)
  if s.size == count 
    sum += s.to_i
    @ans += sum 
  else
    dfs(s,sum,count+1)
    dfs(s[count..-1],sum + s[0..count-1].to_i,1)
  end
end 

puts dfs(s,0,1)

File: arc061/many_formulas.cpp
#include <bits/stdc++.h>
using namespace std;
string s;
long ans=0;

int dfs(long i,long j){
  if(i==s.size()-1){
    return j*10 + (long)(s[i]-'0');
  }else{
    cout << dfs(i+1,j*10+(long)(s[i]-'0')) << endl;
    cout << j*10 + (long)(s[i]-'0')+dfs(i+1,0) << endl;
  }
}

int main() {
  cin >> s;
  dfs(0,0);
  cout << ans << endl;
}